memtable/write_buffer_manager.cc

```diff
@@ -84,26 +117,25 @@ void WriteBufferManager::ReserveMemWithCache(size_t mem) {
 }
 
 void WriteBufferManager::ScheduleFreeMem(size_t mem) {
+  std::unique_lock<std::mutex> lock(wbm_mutex_);
   if (enabled()) {
     memory_active_.fetch_sub(mem, std::memory_order_relaxed);
   }
 }
 
 void WriteBufferManager::FreeMem(size_t mem) {
+  std::unique_lock<std::mutex> lock(wbm_mutex_);
```

Comment:
Can we acquire the lock inside these functions FreeMemWithCache, MaybeEndWriteStall etc to reduce the lock duration?

Reply:
Is there a concrete concern where and how such lock duration can be problematic?  As far as I know, the longest duration come from `cache_res_mgr_->UpdateCacheReservation()` evicting dummy entries linear to `mem` to be freed from cache. Would that be long enough to be problem? 

My intention is to make `FreeMem` an atomic function since read/write to `memory_used_` and `MaybeEndWriteStall()` are very related through `ShouldStall()`. Making these two into one atomic option relieves us from thinking about several data race.

We also have a plan of considering the returned status `s` of `cache_res_mgr_->UpdateCacheReservation()` in `WriteBufferManager::FreeMemWithCache()` (or `ReserveMemWithCache()`)  into the decision of we should stall or not. Now we only do `s.PermitUncheckedError();` So from this perspective, I'd like to make `FreeMemWithCache` atomic with read/write to `memory_used_` and `MaybeEndWriteStall()`.